src/mod/dns.mod/coredns.c

```diff
@@ -1201,6 +1204,103 @@ static void dns_lookup(sockname_t *addr)
   sendrequest(rp, T_PTR);
 }
 
+static int dns_hosts(char *hostn) {
+  #define PATH "/etc/hosts"
+  int fd, len, i, found = 0;
+  struct stat sb;
+  char *addr, hostn_lower[256], hostn_upper[256], *last_newline, *last_hash = NULL, *c, *c2, ip[256];
+  sockname_t name;
+
+  if (!*hostn) {
+    ddebug0(RES_MSG "bogus empty hostname input");
+    return 1;
+  }
+  len = strlen(hostn);
+  if (len > 255) {
+    ddebug0(RES_MSG "bogus len of hostname > 255 input");
+    return 1;
+  }
+  /* due to strncasecmp() and strncmp() being slow if used in loop, precalculate
+   * lower and upper string from hostn and compare with handcrafted code */
+  for (i = 0; i < len; i++) {
+      if (isspace(hostn[i])) {
+        ddebug0(RES_MSG "bogus white-space hostname input");
+        return 1;
+      }
+      hostn_lower[i] = tolower(hostn[i]);
+      hostn_upper[i] = toupper(hostn[i]);
+  }
+  hostn_lower[i] = 0;
+  hostn_upper[i] = 0;
+  fd = open(PATH, O_RDONLY);
+  if (fd < 0) {
+    ddebug1(RES_MSG "open(" PATH "): %s ", strerror(errno));
+    return 0;
+  }
+  if (fstat(fd, &sb) < 0) {
+    ddebug1(RES_MSG "fstat(" PATH "): %s ", strerror(errno));
+    close(fd);
+    return 0;
+  }
+  addr = mmap(NULL, sb.st_size, PROT_READ, MAP_PRIVATE, fd, 0);
+  if (addr == MAP_FAILED) {
+    ddebug1(RES_MSG "mmap(" PATH "): %s ", strerror(errno));
+    close(fd);
+    return 0;
+  }
+  last_newline = addr - 1;
```

Comment:
this calculates a pointer one before an allocated space, it's undefined behavior

Reply:
im sure we can do this "hack" in C. but if you still insist, i will change the code.

Reply:
Fixed pointer arithmetic hack